### DISCUSSION TITLE
ADSR issue

### DISCUSSION THREAD

**Discussion by rolfdegen** [ElectroTechnique/electrotechnique.github.io#42] [Software Issues]
The release time is adjusted automatically when I turn the attack control. Is that right ?

Can someone check that. Thanks :)

**Comment by Electrotechnique** [comment]
Is this the Master branch? Don't use the Multi-timbral branch because we're still working on it and there may be bugs.

**Comment by rolfdegen** [comment]
Thanks.

**Comment by rolfdegen** [comment]
I have integrated the new effect_envelope.h and effect_envelope.cpp into my code. But I have errors when compiling

Warning: 	In member function void AudioEffectEnvelopeTS::release(float)		C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.h	-1
Error:
Line 126:17: error: void AudioEffectEnvelopeTS::release(float) causes a section type conflict with int compare(const void*, const void*)

my code 
FLASHMEM void updateFilterRelease() {
  filterEnvelope1.release(filterRelease);
  filterEnvelope2.release(filterRelease);
  filterEnvelope3.release(filterRelease);
  filterEnvelope4.release(filterRelease);
  filterEnvelope5.release(filterRelease);
  filterEnvelope6.release(filterRelease);
  filterEnvelope7.release(filterRelease);
  filterEnvelope8.release(filterRelease);

**Comment by rolfdegen** [comment]
mmm.. 

i want only new exp envelopes in my code.  i have change only effect_envelope.h and effect_envelope.cpp in my code.

can anyone help me ? Thanks a lot :)

**Comment by rolfdegen** [comment]
OK. I'm testing and removed updateExpRelease() in the effect_envelope.h and I have a linear envelope line and no build errors.

FLASHMEM void release(float milliseconds) {
   release_count = milliseconds2count(milliseconds);
   if (release_count == 0) release_count = 1;
	//updateExpRelease(); }

When I activate updateExpRelease () I get an build error :(

Error:  Build failed for project 'TSynth'
Error: Line  217:16: error: void AudioEffectEnvelopeTS::release(float) causes a section type conflict with int compare(const void*, const void*)		C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.h	217

https://i.ibb.co/rf07fCp/TSError.jpg

**Comment by rolfdegen** [comment]
Big code errors in my code. grrrrrrrrrrrrrrrrr   Why is it so complicated ???????

**Comment by rolfdegen** [comment]
Errrrrrrrrrrrrrrrrrrrrrrror


effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()

Error linking for board Teensy 4.1
effect_envelope.cpp:77: multiple definition of AudioEffectEnvelopeTS  update()
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
ld.exe: Disabling relaxation: it will not work with multiple definitions
Build failed for project 'TSynth'
 
effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()
effect_envelope.cpp:77: multiple definition of AudioEffectEnvelopeTS  noteOn()
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
 
effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()
effect_envelope.cpp:77: multiple definition of AudioEffectEnvelopeTS  noteOff()
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
 
effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()
effect_envelope.cpp:77: multiple definition of AudioEffectEnvelopeTS  isActive()
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
 
effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()
effect_envelope.cpp:77: multiple definition of AudioEffectEnvelopeTS  isSustain()
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
 
effect_envelope.cpp.o: In function AudioEffectEnvelopeTS::update()
effect_envelope.cpp:77: multiple definition of vtable for AudioEffectEnvelopeTS
effect_envelope.cpp.o: C:\Program Files (x86)\Arduino\mySketch\Jeannie102\Jeannie102\effect_envelope.cpp:77: first defined here
